src/interface/variable.cpp

```diff
@@ -39,6 +39,8 @@ Variable<T>::Variable(const std::string &base_name, const Metadata &metadata,
                            "Mismatch between sparse flag and sparse ID");
   uid_ = get_uid_(label());
 
+  pmb_gid_ = wpmb.lock().get()->gid;
+
```

Comment:
It is possible that after a re-mesh when amr is turned on a block can have its `gid` changed. I don't think there is any mechanism to propagate this through to the variables, so I think that the `VariableState` could end up with a stale `gid` in this case. 

The two possible solutions I think of. First, maybe adding a method to `MeshBlock` called `UpdateGid(int gid)` which updates the block `gid` and the `VariableState` of all variables (but I guess this depends on iterating through all of the containers). This method would need to be called in amr_loadbalance.cpp instead of just setting the block `gid`s by hand. The other possibility is to just make sure that the values are updated during sparse packing, which is a little less clean I think but probably easier to implement.

Reply:
Just double checking: are our gids always guaranteed to be contiguous  (starting from 0)?
We have some code paths that implicitly assume that there's always a block with gid 0.

Reply:
Yes, `gid`s are defined to always start at zero and be contiguous (i.e. they are just the indices of all the blocks in a std::vector that has been sorted by the Morton number of the blocks' logical locations). The issue is that after remeshing, the Morton number of a block that hasn't been refined or derefined doesn't change but it's `gid` can change since blocks with lower Morton numbers may have been added/removed from the grid.